### PR TITLE
[MI-3355]: Added dropdown to choose teams and channels

### PR DESCRIPTION
**Pull request by Kshitij-Katiyar**
Developed dropdown with list of all the teams and channels to select.

### CODE REVIEW

File: webapp/src/api/api_wrapper.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import Cookies from 'js-cookie';
+
+import Utils from 'utils';
+
+export const fetchChannels = async (mmSiteUrl: string) => {
+    const authToken = 'Bearer ' + Cookies.get('MMAUTHTOKEN') || '';
```
Comment (raghavaggarwal2308) [inline webapp/src/api/api_wrapper.ts]: Create a util function for this

File: webapp/src/api/api_wrapper.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    const response = await fetch(url, {
+        method: 'GET',
+        headers: {Authentication: authToken},
+    });
+    return response.json();
+};
+
+export const fetchTeams = async (mmSiteUrl: string) => {
+    const authToken = 'Bearer ' + Cookies.get('MMAUTHTOKEN') || '';
+    const url = Utils.getBaseUrls(mmSiteUrl).mattermostApiBaseUrl + '/teams';
```
Comment (raghavaggarwal2308) [inline webapp/src/api/api_wrapper.ts:13]: Contants for these API paths

File: webapp/src/components/modals/actionModal.tsx [hunk 3]
```diff
@@ -0,0 +1,173 @@
+import React, {useEffect, useState} from 'react';
+
+import Button from 'react-bootstrap/Button';
+import Modal from 'react-bootstrap/Modal';
+import Table from 'react-bootstrap/Table';
+import Form from 'react-bootstrap/Form';
+
+import './styles.css';
+
+import {OverlayTrigger, Tooltip} from 'react-bootstrap';
```
Comment (raghavaggarwal2308) [inline webapp/src/components/modals/actionModal.tsx:7]: Group it with the other react-bootstrap imports

Comment (raghavaggarwal2308) [reply webapp/src/components/modals/actionModal.tsx:7]: I think you can add all the other imports in this import only. No need to import them separately.

File: webapp/src/components/modals/actionModal.tsx [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    config: Configs[];
+    configIndex: number;
+}
+
+const ActionModal = ({visibility, setVisibility, config, configIndex}: Props) => {
+    const [show, setShow] = useState(false);
+
+    const [attachmentMessageAvailable, setAttachmentMessageAvailable] = useState(false);
+
+    const actionsLength = config[configIndex]?.actions?.length ?? 0;
```
Comment (raghavaggarwal2308) [inline webapp/src/components/modals/actionModal.tsx]:
```suggestion
    const actionsLength = config[configIndex]?.actions?.length;
```
I think this will work!

File: webapp/src/components/modals/actionModal.tsx [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+    const checkAttachmentMessage = () => {
+        if (config[configIndex]?.attachmentMessage) {
+            if (config[configIndex]?.attachmentMessage?.length === 0) {
+                setAttachmentMessageAvailable(false);
+            } else {
+                setAttachmentMessageAvailable(config[configIndex]?.attachmentMessage?.[0] !== '');
+            }
+        }
+    };
```
Comment (raghavaggarwal2308) [inline webapp/src/components/modals/actionModal.tsx]:
```suggestion
    const checkAttachmentMessage = () => {
        if (config[configIndex]?.attachmentMessage?.length) {
            setAttachmentMessageAvailable(Boolean(config[configIndex]?.attachmentMessage?.[0]));
            return
        }
        
        setAttachmentMessageAvailable(false);
    };
```

File: webapp/src/components/modals/actionModal.tsx [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
+    useEffect(() => {
+        checkAttachmentMessage();
+        setShow(visibility);
+    }, [visibility]);
+
+    const handleClose = () => {
+        setShow(false);
+        setVisibility(false);
+    };
```
Comment (raghavaggarwal2308) [inline webapp/src/components/modals/actionModal.tsx]:
```suggestion
    };
    
```

File: webapp/src/components/modals/actionModal.tsx [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+                className='customModal'
+                show={show}
+                onHide={handleClose}
+            >
+                <Modal.Header closeButton={false}>
+                    <Modal.Title>{'Actions'}</Modal.Title>
+                </Modal.Header>
+
+                <Modal.Body className='customModalBody'>
+                    {attachmentMessageAvailable || (config[configIndex]?.actions && actionsLength > 0) ? (<>
```
Comment (raghavaggarwal2308) [inline webapp/src/components/modals/actionModal.tsx]:
```suggestion
                    {attachmentMessageAvailable || (config[configIndex]?.actions && actionsLength) ? (<>
```

File: webapp/src/components/modals/actionModal.tsx [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+                                        value={config[configIndex].attachmentMessage ?? ''}
+                                        placeholder=''
+                                        aria-label='Disabled input example'
+                                        readOnly={true}
+                                    />
+                                </Form.Group>
+                            </Form>
+                        ) : (<p>{'No attachment message configured'}</p>)
+                        }
+                        {config[configIndex]?.actions && actionsLength > 0 ? (
```
Comment (raghavaggarwal2308) [inline webapp/src/components/modals/actionModal.tsx]: same

File: webapp/src/components/modals/actionModal.tsx [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+                            <div>
+                                <Form>
+                                    <Form.Group className='action-group'>
+                                        <Form.Label>{'Actions'}</Form.Label>
+                                    </Form.Group>
+                                </Form>
+                                <div className='listTable'>
+                                    <Table
+                                        striped={true}
+                                        className='listTable'
```
Comment (raghavaggarwal2308) [inline webapp/src/components/modals/actionModal.tsx]: Parent and child have the same className. Is this what you want to do?

File: webapp/src/components/modals/actionModal.tsx [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+                                    >
+                                        <thead>
+                                            <tr>
+                                                <th className='type-action'>{'Type'}</th>
+                                                <th className='display-name-action'>{'Display Name'}</th>
+                                                <th className='name-action'>{'Name'}</th>
+                                                <th className='channels-added-action'>{'Add to Channels'}</th>
+                                                <th className='successfull-message-action'>{'Success Message'}</th>
```
Comment (raghavaggarwal2308) [inline webapp/src/components/modals/actionModal.tsx]:
```suggestion
                                                <th className='success-message-action'>{'Success Message'}</th>
```

File: webapp/src/components/modals/actionModal.tsx [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+import Table from 'react-bootstrap/Table';
+import Form from 'react-bootstrap/Form';
+
+import './styles.css';
+
+import {OverlayTrigger, Tooltip} from 'react-bootstrap';
+
+type Props = {
+    visibility: boolean;
+    setVisibility: React.Dispatch<React.SetStateAction<boolean>>;
```
Comment (raghavaggarwal2308) [inline webapp/src/components/modals/actionModal.tsx:13]: I don't see any use of these in the component.

Comment (raghavaggarwal2308) [reply webapp/src/components/modals/actionModal.tsx:13]: Check if these can be removed

File: webapp/src/components/modals/configModal.tsx [hunk 12]
```diff
@@ -0,0 +1,739 @@
+import React, {useEffect, useState} from 'react';
+
+import Button from 'react-bootstrap/Button';
+import Modal from 'react-bootstrap/Modal';
+import Form from 'react-bootstrap/Form';
+import Table from 'react-bootstrap/Table';
+import ButtonGroup from 'react-bootstrap/ButtonGroup';
+import {OverlayTrigger, Tooltip, ToggleButton} from 'react-bootstrap';
```
Comment (raghavaggarwal2308) [inline webapp/src/components/modals/configModal.tsx:8]: Check if these can be combined in a single import

File: webapp/src/components/modals/configModal.tsx [hunk 13]
```diff
@@ -0,0 +1,739 @@
...
+
+import {useSelector} from 'react-redux';
+
+import {GlobalState} from 'mattermost-redux/types/store';
+
+import './styles.css';
+
+import {fetchChannelsAndTeams} from 'api/api_wrapper';
+
+import {DeleteSvg, EditSvg} from '../svgIcons/svg';
```
Comment (raghavaggarwal2308) [inline webapp/src/components/modals/configModal.tsx:18]: Fix import order in this file

File: webapp/src/components/modals/configModal.tsx [hunk 14]
```diff
@@ -0,0 +1,739 @@
...
+
+type Props = {
+    visibility: boolean;
+    setVisibility: React.Dispatch<React.SetStateAction<boolean>>;
+    configIndex: number | null;
+    config: Configs[];
+    onChange: any;
```
Comment (raghavaggarwal2308) [inline webapp/src/components/modals/configModal.tsx]: Use proper types

File: webapp/src/components/modals/configModal.tsx [hunk 15]
```diff
@@ -0,0 +1,739 @@
...
+
+    const actionElement: Actions = {
+        actionType: '',
+        actionName: '',
+        actionDisplayName: '',
+        channelsAddedTo: [''],
+        actionSuccessfullMessage: [''],
+    };
+    const newAction: Actions[] = [
+    ];
```
Comment (raghavaggarwal2308) [inline webapp/src/components/modals/configModal.tsx]: single line

File: webapp/src/components/modals/configModal.tsx [hunk 16]
```diff
@@ -0,0 +1,739 @@
...
+                    <Button
+                        variant='primary'
+                        onClick={handlePrimary}
+                    >
+                        {'Save changes'}
+                    </Button>
+                </Modal.Footer>
+            </Modal>
+        </div>
+
```
Comment (raghavaggarwal2308) [inline webapp/src/components/modals/configModal.tsx]:
```suggestion
```

File: webapp/src/components/modals/configModal.tsx [hunk 17]
```diff
@@ -0,0 +1,739 @@
...
+                            </div>
+                        </Form>
+                    </div>}
+                    {!isDeleteVisible && !isActionVisible && <div className='add-action-button'>
+                        <Button
+                            className={actionLength > 0 ? 'add-actions' : ''}
+                            onClick={handleAddActions}
+                        >{'Add actions'}</Button>
+                    </div>}
+
```
Comment (raghavaggarwal2308) [inline webapp/src/components/modals/configModal.tsx]:
```suggestion
```

File: webapp/src/components/modals/deleteModal.tsx [hunk 18]
```diff
@@ -0,0 +1,62 @@
+import React, {useEffect, useState} from 'react';
+
+import Button from 'react-bootstrap/Button';
+import Modal from 'react-bootstrap/Modal';
+
+import './styles.css';
+
+type Props = {
+    visibility: boolean;
+    setVisibility: React.Dispatch<React.SetStateAction<boolean>>;
```
Comment (raghavaggarwal2308) [inline webapp/src/components/modals/deleteModal.tsx:10]: same

File: webapp/src/components/tables/styles.css [hunk 19]
```diff
@@ -40,15 +48,15 @@
 
 .existing-config-table .btn-group.btn.btn-default:hover, .btn-group .btn.btn-default:focus {
     background: transparent;
-    color: #000;
+    color: #000000;
```
Comment (raghavaggarwal2308) [inline webapp/src/components/tables/styles.css]: why this change

**Review by raghavaggarwal2308 — CHANGES_REQUESTED** [review]
If the CSS for all three modals is different, you should separate each modal in a folder and their CSS should be in that folder.

File: webapp/src/components/modals/actionModal.tsx (hunk 3 above)
Comment (Kshitij-Katiyar) [reply webapp/src/components/modals/actionModal.tsx:7]: I have tried importing all the bootstrap component in this import only. But it is giving random errors. This was the fix suggested by internet

File: webapp/src/components/modals/actionModal.tsx (hunk 11 above)
Comment (Kshitij-Katiyar) [reply webapp/src/components/modals/actionModal.tsx:13]: it is used to hide the modal when close button in clicked, i don't think we can remove it

File: webapp/src/components/modals/configModal.tsx (hunk 12 above)
Comment (Kshitij-Katiyar) [reply webapp/src/components/modals/configModal.tsx:8]: it is not working in single import

File: webapp/src/components/modals/actionModal.tsx (hunk 9 above)
Comment (Kshitij-Katiyar) [reply webapp/src/components/modals/actionModal.tsx]: no, fixed it

File: webapp/src/components/modals/deleteModal.tsx (hunk 18 above)
Comment (Kshitij-Katiyar) [reply webapp/src/components/modals/deleteModal.tsx:10]: no, we cannot remove this

File: webapp/src/components/tables/styles.css (hunk 19 above)
Comment (Kshitij-Katiyar) [reply webapp/src/components/tables/styles.css]: should we use #000 for black or #000000 ?

Comment (raghavaggarwal2308) [reply webapp/src/components/tables/styles.css]: #000

File: webapp/src/api/api_wrapper.ts [hunk 20]
```diff
@@ -0,0 +1,22 @@
+import Utils from 'utils';
+
+export const fetchChannels = async (mmSiteUrl: string) => {
+    const authToken = Utils.getAuthToken();
+    const url = Utils.getBaseUrls(mmSiteUrl).mattermostApiBaseUrl + '/channels?exclude_default_channels=true';
+    const response = await fetch(url, {
+        method: 'GET',
+        headers: {Authentication: authToken},
```
Comment (raghavaggarwal2308) [inline webapp/src/api/api_wrapper.ts]:
```suggestion
        headers: {Authentication: Utils.getAuthToken()},
```
Remove the authToken variable

File: webapp/src/api/api_wrapper.ts [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+    });
+    return response.json();
+};
+
+export const fetchTeams = async (mmSiteUrl: string) => {
+    const authToken = Utils.getAuthToken();
```
Comment (raghavaggarwal2308) [inline webapp/src/api/api_wrapper.ts]: same

File: webapp/src/components/modals/actionModal.tsx (hunk 3 above)
Comment (raghavaggarwal2308) [reply webapp/src/components/modals/actionModal.tsx:7]: Group the imports

File: webapp/src/components/modals/styles.css [hunk 22]
```diff
@@ -0,0 +1,284 @@
...
+.config-form {
+  width: 100%;
+}
+
+.add-action-button {
+  display: block;
+}
+
+.guestButton.active {
+  box-shadow: none;
```
Comment (raghavaggarwal2308) [inline webapp/src/components/modals/styles.css]: Isn't this by default?

**Review by raghavaggarwal2308 — CHANGES_REQUESTED** [review]

File: webapp/src/components/modals/styles.css (hunk 22 above)
Comment (Kshitij-Katiyar) [reply webapp/src/components/modals/styles.css]: no, radio buttons have a shadow by default

**Review by raghavaggarwal2308 — APPROVED** [review]

File: webapp/src/components/modals/actionModal.tsx [hunk 23]
```diff
@@ -0,0 +1,167 @@
...
+        setAttachmentMessageAvailable(false);
+    };
+
+    useEffect(() => {
+        checkAttachmentMessage();
+    }, []);
+
+    const handleClose = () => {
+        setVisibility(false);
+    };
```
Comment (avas27JTG) [inline webapp/src/components/modals/actionModal.tsx]:
```suggestion
    const handleClose = () => setVisibility(false);
```

File: webapp/src/components/modals/actionModal.tsx [hunk 24]
```diff
@@ -0,0 +1,167 @@
...
+
+    return (
+        <div>
```
Comment (avas27JTG) [inline webapp/src/components/modals/actionModal.tsx]: remove if not needed or use `<> </>`

File: webapp/src/components/modals/styles.scss [hunk 25]
```diff
@@ -0,0 +1,280 @@
...
+	margin-bottom: 25px;
+}
+
+.form-group {
+	position: relative;
+	margin-bottom: 10px;
+}
+
+.warnings {
+	margin-bottom: 25.15px;
```
Comment (avas27JTG) [inline webapp/src/components/modals/styles.scss]: avoid such decimal considerations see if `25` or `26` works

File: webapp/src/components/modals/styles.scss [hunk 26]
```diff
@@ -0,0 +1,280 @@
...
+.name-action {
+	width: 15%;
+}
+
+.display-name-action {
+	width: 21%;
+}
+
+.option-buttons {
+	width: 17%;
```
Comment (avas27JTG) [inline webapp/src/components/modals/styles.scss]: use `px` `%` should not be used this way

File: webapp/src/components/svgIcons/svg.tsx [hunk 27]
```diff
@@ -0,0 +1,79 @@
...
+            stroke='#333333'
+            strokeWidth='1.65'
+            strokeLinecap='round'
+            strokeLinejoin='round'
+        >
+            <path d='M20 14.66V20a2 2 0 0 1-2 2H4a2 2 0 0 1-2-2V6a2 2 0 0 1 2-2h5.34'/>
+            <polygon points='18 2 22 6 12 16 8 16 8 12 18 2'/>
+        </svg>
+    );
+};
```
Comment (avas27JTG) [inline webapp/src/components/svgIcons/svg.tsx]:
```suggestion
const EditSvg = () => (
      <svg
            className='svg'
            xmlns='http://www.w3.org/2000/svg'
            width='16'
            height='16'
            viewBox='0 0 24 24'
            fill='none'
            stroke='#333333'
            strokeWidth='1.65'
            strokeLinecap='round'
            strokeLinejoin='round'
        >
            <path d='M20 14.66V20a2 2 0 0 1-2 2H4a2 2 0 0 1-2-2V6a2 2 0 0 1 2-2h5.34'/>
            <polygon points='18 2 22 6 12 16 8 16 8 12 18 2'/>
        </svg>
    );
```

File: webapp/src/components/svgIcons/svg.tsx [hunk 28]
```diff
@@ -0,0 +1,79 @@
...
+            <line
+                x1='14'
+                y1='11'
+                x2='14'
+                y2='17'
+            />
+        </svg>
+
+    );
+};
```
Comment (avas27JTG) [inline webapp/src/components/svgIcons/svg.tsx]: same

File: webapp/src/components/svgIcons/svg.tsx [hunk 29]
```diff
@@ -0,0 +1,79 @@
...
+            <path d='M1 12s4-8 11-8 11 8 11 8-4 8-11 8-11-8-11-8z'/>
+            <circle
+                cx='12'
+                cy='12'
+                r='3'
+            />
+        </svg>
+    );
+};
+
```
Comment (avas27JTG) [inline webapp/src/components/svgIcons/svg.tsx]: same

File: webapp/src/components/tables/existingConfigTable.tsx [hunk 30]
```diff
@@ -0,0 +1,189 @@
...
+    };
+
+    const handleEdit = (index: number) => {
+        setConfigIndex(index);
+        setIsEditVisible(true);
+    };
+
+    const handleAdd = () => {
+        setIsAddVisible(true);
+    };
```
Comment (avas27JTG) [inline webapp/src/components/tables/existingConfigTable.tsx]:
```suggestion
    const handleAdd = () =>  setIsAddVisible(true);
```
fix all these

for arrow function there is no need for explicit return if there is only statement in the function body

File: webapp/src/components/modals/styles.scss (hunk 25 above)
Comment (Kshitij-Katiyar) [reply webapp/src/components/modals/styles.scss]: used 25px

**Review by avas27JTG — APPROVED** [review]
